### PR TITLE
Add .editorconfig

### LINKED ISSUE #3209: Add .editorconfig
Perhaps we should add a config file for the line endings and stuff

http://editorconfig.org/

Ref #3200
Ref #3197

### PR DESCRIPTION
**Pull request by MattBlissett**
Fixes #3209 

This should configure editors to follow the indentation and whitespace behaviour the existing code has, using http://editorconfig.org/

### CODE REVIEW

**Comment by matthijsmelissen** [comment]
Looks good, thanks!

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# http://EditorConfig.org
+
+root = true
+
+# Defaults apply to all files
+[*]
+charset = utf-8
+end_of_line = lf
+insert_final_newline = true
+trim_trailing_whitespace = true
```
Comment (matthijsmelissen) [inline .editorconfig]: This one I'm not sure about: obviously we don't want trailing whitespace, but we also don't want commits to include trimming of trailing whitespace in unrelated code. Such trailing whitespace might be introduced accidentally by users with editors not honoring .editconfig.

What do you think is best?

Comment (MattBlissett) [reply .editorconfig]: I have `trim_trailing_whitespace = true` for all my work projects, and haven't had any problems — but I'm mostly working with people using IDEs or editors that have built-in support for Editorconfig.

One other option is a Git pre-commit hook, but this [looks complicated](https://github.com/imoldman/config/blob/master/pre-commit.git.sh) so I won't recommend it.

Comment (matthijsmelissen) [reply .editorconfig]: Another option would be a travis check that fails if people add trailing whitespace.

Comment (kocio-pl) [reply .editorconfig]: I prefer Travis. Trimming is turned on by default in Atom for example and it caused some PRs to be cluttered with such small fixes not related to the subject.

Comment (MattBlissett) [reply .editorconfig]: If Travis were set to fail because of trailing whitespace, then adding the same rule in Editorconfig is convenient for users whose editors support it.

At present, the only files with trailing whitespace are Markdown — they could be excluded from this rule.

**Comment by kocio-pl** [comment]
Maybe any kind of pushing people to remove trailing spaces is bad, since automatic removing them in the background creates mess and Travis will not make them create separate patches anyway. Since this problem is quite small and simple solutions make more harm than good, I would rather prefer that some checking and cleaning would be run periodically.

**Comment by matthijsmelissen** [comment]
Thanks!
